Add support to xml data type

suarezrominajulieta opened the pull request.

arcuri82 reviewed (changes requested):
@suarezrominajulieta  thx for your first PR! see comments

suarezrominajulieta commented:
thank you for the comments! I'll look into them

suarezrominajulieta commented on core/src/main/kotlin/org/evomaster/core/search/gene/ObjectGene.kt:
```diff
@@ -304,6 +309,99 @@ class ObjectGene(
+            .replace("\"", "&quot;")
+            .replace("'", "&apos;")
+
+    private fun singularize(n: String): String =
```
Done, used the reference of the childs objects instead.

suarezrominajulieta commented on core/src/main/kotlin/org/evomaster/core/search/gene/ObjectGene.kt:
```diff
@@ -304,6 +309,99 @@ class ObjectGene(
+    public fun cleanXmlValueString(v: String): String =
+        v.removeSurrounding("\"").let(::escapeXmlSafe)
+
+    private fun getPrintedValue(
```
Fixed and absorved

suarezrominajulieta commented on core/src/main/kotlin/org/evomaster/core/search/gene/ObjectGene.kt:
```diff
@@ -304,6 +309,99 @@ class ObjectGene(
+
+            is Gene -> "<$name>${getPrintedValue(previousGenes, v, targetFormat)}</$name>"
+
+            else -> "<$name>${cleanXmlValueString(v.toString())}</$name>"
```
We are not, fixed

suarezrominajulieta commented on core/src/main/kotlin/org/evomaster/core/problem/rest/builder/RestActionBuilderV3.kt:
```diff
@@ -748,8 +747,12 @@ object RestActionBuilderV3 {
         }
 
-        var gene = getGene("body", obj.schema, schemaHolder,currentSchema, referenceClassDef = null, options = options, messages = messages, examples = examples)
+        val deref = obj.schema.`$ref`?.let { ref -> val name = ref.substringAfterLast("/")
```
Done, I've improved identation and removed the unused "name" variable.

suarezrominajulieta commented on core/src/main/kotlin/org/evomaster/core/search/gene/ObjectGene.kt:
```diff
@@ -304,6 +309,99 @@ class ObjectGene(
+            )
+        )
+
+    private fun isPrimitiveGene(value: Any?): Boolean =
```
Done, i verified if the object had childs instead.

suarezrominajulieta commented on core/src/main/kotlin/org/evomaster/core/search/gene/ObjectWithAttributesGene.kt:
```diff
@@ -0,0 +1,128 @@
+            .keys
+
+        if (duplicated.isNotEmpty()) {
+            throw IllegalStateException("Duplicate child elements not allowed in XML: $duplicated")
```
Done: i moved the validation to the constructor, and added logs to restBuilderAction

suarezrominajulieta commented on core/src/test/kotlin/org/evomaster/core/search/gene/ObjectGeneTest.kt:
```diff
@@ -101,4 +101,40 @@ internal class ObjectGeneTest {
 
         assertEquals("{foo,bar,nested{hello}}", actual)
     }
+
```
Done. I've added:
- a White-box E2E test in spring-rest-openapi-v3: XMLEMTest
- a Black-box E2E test in spring-rest-bb: BBXMLTest

arcuri82 commented on core/src/test/kotlin/org/evomaster/core/search/gene/GeneRandomizedTest.kt:
```diff
@@ -74,7 +75,39 @@ class GeneRandomizedTest : AbstractGeneTest(){
+                val containsSame = x.containsSameValueAs(y)
+                if (containsSame) {
+                    // Debug info: find which child gene has the inconsistency
+                    val debugInfo = buildString {
```
this could be refactored to a call to a private method

suarezrominajulieta replied:
i'm Sorry, this file was not supposed to be modified.

arcuri82 commented on core/src/test/kotlin/org/evomaster/core/search/gene/GeneRandomizedTest.kt:
```diff
@@ -74,7 +75,39 @@ class GeneRandomizedTest : AbstractGeneTest(){
+                        }
+                        appendLine("=== END DEBUG ===")
+                    }
+                    System.err.println(debugInfo)
```
this `debugInfo` should not go on `System.err.println`, but rather appended to the error message of the `assertFalse`

suarezrominajulieta replied:
i'm Sorry, this file was not supposed to be modified

suarezrominajulieta commented:
Hi, i made the changes requested: i fixed the bbxml test, add the comments, removed the spanish comments, and the changes to the GeneRandomisedTest, that i didn't want to change, it was a mistake. Thank you!

suarezrominajulieta commented on core/src/main/kotlin/org/evomaster/core/search/gene/ObjectWithAttributesGene.kt:
```diff
@@ -0,0 +1,160 @@
+        // If other is also ObjectWithAttributesGene, attributeNames must match
+        // If other is plain ObjectGene, this.attributeNames must be empty to produce same XML
+        if (other is ObjectWithAttributesGene) {
+            if (this.attributeNames != other.attributeNames) {
```
Hi @arcuri82 — correct me if I'm wrong, but in Kotlin != calls .equals() under the hood, so this is comparing by content, not by reference (reference identity would require !==). Since Set.equals() checks that both sets contain the same elements regardless of order, I believe the comparison is correct: two ObjectWithAttributesGene instances with the same attribute names — even if constructed separately — should be considered equivalent for the purposes of this method. Please let me know if you had something else in mind!
